### ISSUE TITLE
uninlined_format_args is too strict around renamed args

### ISSUE THREAD

**Issue by conradludgate** [rust-lang/rust-clippy#10247]
### Summary

clippy suggests inlining a format arg that has been explicitly renamed

### Lint Name

uninlined_format_args

### Reproducer

I tried this code:

```rust
fn main() {
    let val = "World";
    println!("Hello, {name}!", name = val);
}
```

I saw this happen:

```
warning: variables can be used directly in the `format!` string
 --> src/main.rs:3:5
  |
3 |     println!("Hello, {name}!", name = val);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args
  = note: `#[warn(clippy::uninlined_format_args)]` on by default
help: change this to
  |
3 -     println!("Hello, {name}!", name = val);
3 +     println!("Hello, {val}!");
  |
```

I expected to see this happen:

No warning

### Version

```text
rustc 1.67.0 (fc594f156 2023-01-24)
binary: rustc
commit-hash: fc594f15669680fa70d255faec3ca3fb507c3405
commit-date: 2023-01-24
host: aarch64-apple-darwin
release: 1.67.0
LLVM version: 15.0.6
```

**Comment by lukaslueg** [comment]
This case is correctly handled afaics:

```raw
help: change this to
  |
3 -     println!("Hello, {name}!", name = val);
3 +     println!("Hello, {val}!");
  |
```

so it's not exactly a false positive. The question is if an explicit rename that could be removed should get linted.

**Comment by Manishearth** [comment]
cc @nyurik

**Comment by nyurik** [comment]
This was intentional - `name` in this case is not a variable but an alias to a variable -- and when inlining, the alias is no longer needed. This way it also solves the common case of `format!("{a} {b} {c}", a=something, b=else, c=with_long_names)`, regardless if the a,b,c are the same as the vars or not.

**Comment by conradludgate** [comment]
I disagree with the decision here. I believe that aliases are useful and can be used intentionally.

> This way it also solves the common case of format!("{a} {b} {c}", a=something, b=else, c=with_long_names)

Is this common? I've not seen this use of aliases in the wild

(also, I'm happy to move this discussion elsewhere if this false-positive is not the correct place, this is the best issue template I could come up with to describe my thoughts on this matter)

**Comment by Manishearth** [comment]
Yeah, I'd also say that aliases might be used intentionally, to make the string have a clear "format". This might just be a case of "apply `allow()` when you're doing that",I've definitely used aliases that way, but not _often_.


I kinda feel like for this lint it makes sense to do the replacement if people are using positional args or `foo=foo` args, but I think it's reasonable to skip this case.

> also, I'm happy to move this discussion elsewhere if this false-positive is not the correct place

The discussion is in the correct place, though false positive is not quite the right categorization (close enough, though)

> I disagree with the decision here

(no decision has been made)

**Comment by Manishearth** [comment]
We already have `allow_mixed_uninlined_format_args` as a config option. Should we also make this configurable?

**Comment by nyurik** [comment]
_Someone with rights, please remove the false positive and the bug tags, and lets convert this into a suggestion/feature request?_

I wonder how common is this case?  I could imagine a giant format statement that aliases variable names for the purposes of clarity, but it still seems strange because we are talking about a very small scope within a single function that uses two different names for a variable/param name, and hard-coded format string. Seems we have these options:

#### Use `allow` attribute
Leave as is, and require users to add `allow` whenever this is needed. This path is good if this case is extremely rare per above.  Do we have any way to evaluate how common this is?

#### Ignore when alias name != var name
Not ideal because it won't catch typos or old cases where aliases were done to reduce the verbosity of the format string. Plus anyone who wants inlining in this case too, won't be able to use it.

#### Introduce a new config option
This is ok if there are a lot of such cases, but we would still need to decide which case is more common - the allow or the warn case. The cost of an extra config option should be justifiable - there must be enough of such cases to introduce it as oppose to require the `allow` attrib.
